### PR TITLE
Support relaxed acceptance for thinking tokens in speculative decoding

### PR DESCRIPTION
**Pull request by pyc96**
## Motivation

Support relax the criteria of draft token acceptance if the tokens are part of thinking content. While it will make the accepted tokens distribution shifted from the target model distribution, the hope is thinking tokens changes will not impact the final output quality that much.

## Modifications

- Kernel change to support per sample speculative_accept_threshold_single and speculative_accept_threshold_acc.
- Detect tokens are part of thinking content

## Test

Tested with sharegpt
```
python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-R1-0528 --trust-remote-code --tp 8  --speculative-algorithm=EAGLE --speculative-num-steps=3 --speculative-eagle-topk=1 --speculative-num-draft-tokens=4 --speculative-accept-threshold-acc=0.01 --speculative-accept-threshold-single=0.01 --speculative-relaxed-thinking  --speculative-reasoning-parser=deepseek-r1

python3 -m sglang.bench_serving --backend sglang --num-prompt 10 --max-concurrency 1 --backend=sglang-oai-chat  --temperature=0.6  --apply-chat-template --chat-completions-unset-output-len
```

Baseline:
```
python3 -m sglang.bench_serving --backend sglang --num-prompt 10 --max-concurrency 1 --backend=sglang-oai  --temperature=0.6
```

| | Relax    | Regular |
| - | -------- | ------- |
| accept rate | 2.9 |  2.4 |
| Output throughput @ concurrent = 1 | 140.7 |  120   |

Quality wise, only tested GSM8k, but seems to be on par. (Will need to tune the threshold to find a better one to balance quality and speed. Currently the value seems to be too relaxed)

```
Relaxed:
local-chat-completions (model=nvidia/DeepSeek-R1-0528-FP4,base_url=http://127.0.0.1:30000/v1/chat/completions,num_concurrent=128,timeout=999999), gen_kwargs: (max_gen_toks=16000,temperature=0.6,until=<|end_of_setense|>), limit: None, num_fewshot: None, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9598|±  |0.0054|
|     |       |strict-match    |     5|exact_match|↑  |0.7604|±  |0.0118|

Baseline:
local-chat-completions (model=nvidia/DeepSeek-R1-0528-FP4,base_url=http://127.0.0.1:30000/v1/chat/completions,num_concurrent=128,timeout=999999), gen_kwargs: (max_gen_toks=16000,temperature=0.6,until=<|end_of_setense|>), limit: None, num_fewshot: None, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9644|±  |0.0051|
|     |       |strict-match    |     5|exact_match|↑  |0.7930|±  |0.0112|
```

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: sgl-kernel/csrc/common_extension.cc [hunk 1]
```diff
@@ -221,7 +221,7 @@ TORCH_LIBRARY_FRAGMENT(sgl_kernel, m) {
       "tree_speculative_sampling_target_only(Tensor! predicts, Tensor! accept_index, Tensor! accept_token_num, "
       "Tensor candidates, Tensor retrive_index, Tensor retrive_next_token, Tensor retrive_next_sibling, "
       "Tensor uniform_samples, Tensor uniform_samples_for_final_sampling, Tensor target_probs, Tensor draft_probs, "
-      "float threshold_single, float threshold_acc, "
+      "Tensor threshold_single, Tensor threshold_acc, "
```
Comment (ispobock) [inline sgl-kernel/csrc/common_extension.cc]: Could you put kernel change in a separate PR? We need to update the kernel first -> release new kernel version -> update python code to apply new kernel.

Comment (pyc96) [reply sgl-kernel/csrc/common_extension.cc]: Done, please take a look #8068. Thanks

File: python/sglang/srt/managers/scheduler.py [hunk 2]
```diff
@@ -344,18 +344,39 @@ def __init__(
...
 
+            self.relaxed_thinking = global_server_args_dict.get(
+                "speculative_relaxed_thinking", False
+            )
+            if self.server_args.speculative_reasoning_parser and self.tokenizer:
+                spec_reasoning_parser = ReasoningParser(
+                    model_type=self.server_args.speculative_reasoning_parser,
+                    stream_reasoning=False,
+                )
+                self.think_start_token = self.tokenizer.encode(
```
Comment (CatherineSue) [inline python/sglang/srt/managers/scheduler.py]: nit: is `think_start_token_id` a better name here?

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -531,6 +533,11 @@ def __post_init__(self):
             # If sepculative_num_steps >= speculative_num_draft_tokens, the additional tokens will definitely be discarded.
             # assert self.speculative_num_steps < self.speculative_num_draft_tokens
 
+            if self.speculative_relaxed_thinking:
+                assert (
+                    self.speculative_reasoning_parser
+                ), "Must set reasoning parser when using relaxed thinking."
```
Comment (CatherineSue) [inline python/sglang/srt/server_args.py]: nit: Must set speculative reasoning parser?

File: python/sglang/srt/server_args.py [hunk 4]
```diff
@@ -1253,6 +1260,19 @@ def add_cli_args(parser: argparse.ArgumentParser):
             help="The accept probability of a draft token is raised from its target probability p to min(1, p / threshold_acc).",
             default=ServerArgs.speculative_accept_threshold_acc,
         )
+        parser.add_argument(
+            "--speculative-relaxed-thinking",
+            action="store_true",
+            default=ServerArgs.speculative_relaxed_thinking,
+            help="Relaxed acceptance for thinking tokens.",
+        )
+        parser.add_argument(
```
Comment (CatherineSue) [inline python/sglang/srt/server_args.py:1285]: qq: is speculative-reasoning-parser usually the same as reasoning-parser if set?

Comment (pyc96) [reply python/sglang/srt/server_args.py:1285]: Yep I think so. I added a new flag just to decouple relaxed thinking from using reasoning_content in chat completion API. Since if speculative-reasoning-parser is set, the server will always set reasoning_content. Having 2 flags we can support relaxed thinking but not using reasoning_content

File: python/sglang/srt/server_args.py (hunk 3 above)
Comment (pyc96) [reply python/sglang/srt/server_args.py]: Done

File: python/sglang/srt/managers/scheduler.py (hunk 2 above)
Comment (pyc96) [reply python/sglang/srt/managers/scheduler.py]: Done!

**Comment by yuyu5333** [comment]
@pyc96 Thank you for the author's efforts, but I encountered the following issues when trying to replicate the practice.

update_thinking_states depends on EagleVerifyOutput. In the current commit, there is no python/sglang/srt/speculative/eagle_info.py file, and EagleVerifyOutput does not have the thinking_states parameter:

``` python
# python/sglang/srt/speculative/eagle_worker.py Line 719
res: EagleVerifyOutput = spec_info.verify(
    batch,
    logits_output,
    self.token_to_kv_pool_allocator,
    self.page_size,
    vocab_mask,
)

batch.update_thinking_states(res.thinking_states)
```

Therefore, when I was trying, I removed this line of code...
